File: strings/prob804.py
# 804 Unique Morse Code Words

# 80% solution
class Solution(object):
    def uniqueMorseRepresentations(self, words):
        """
        :type words: List[str]
        :rtype: int
        """
        codes = dict(zip(string.ascii_lowercase,
                   [".-","-...","-.-.","-..",".","..-.","--.","....","..",".---","-.-",".-..","--","-.","---",".--.","--.-",".-.","...","-","..-","...-",".--","-..-","-.--","--.."]))

        words = set(words)

        trans = set()
        for word in words:
            tran = "".join(map(lambda c: codes[c], word))
            trans.add(tran)
        return len(trans)

File: strings/prob804.kt
// 804 Unique Morse Code Words
//

class Solution {
    fun uniqueMorseRepresentations(words: Array<String>): Int {

        var code = listOf(".-","-...","-.-.","-..",".","..-.","--.","....","..",".---","-.-",".-..","--","-.","---",".--.","--.-",".-.","...","-","..-","...-",".--","-..-","-.--","--..")

        var s = words.toSet()
        var ans = mutableSetOf<String>()
        for (w in s) {
            var sb = StringBuilder()
            for (c in w) {
                sb.append(code[c - 'a'])
            }
            ans.add(sb.toString())
        }

        return ans.size
    }
}
